src/core/SampleLoader.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ * SampleLoader.cpp - Sample loader with an optional cache
+ *
+ * Copyright (c) 2023 saker <sakertooth@gmail.com>
+ *               2024 Dalton Messmer <messmer.dalton/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleLoader.h"
+
+#include <QMessageBox>
+#include <cassert>
+
+#include "Engine.h"
+#include "GuiApplication.h"
+#include "PathUtil.h"
+#include "Song.h"
+#include "lmms_basics.h"
+
+namespace lmms {
+
+namespace {
+
+void displayError(const QString& message)
+{
+	Engine::getSong()->collectError(QString("%1: %2").arg(QObject::tr("Error loading sample"), message));
+}
+
+} // namespace
+
+/**
+ * An evictor which automatically evicts sample buffers from the cache
+ * immediately after the last reference to the buffer is dropped
+*/
```

Comment:
This documentation answers my question as to what `AutoEvictor` is. Documentation should ideally be in the header where everybody can see it. Also, the name IMO doesn't give the idea away of it being a deleter quite nicely. Maybe this should be changed?

Reply:
It's more of an implementation detail of the cache - not anything anybody should have to know about. I just declared it in the header so that it could call private `SampleLoader` methods.